packs/BP/scripts/network.ts

```diff
@@ -102,19 +102,36 @@ export class MachineNetwork extends DestroyableObject {
 
     // Calculate the amount of each type that is available to send around.
     const distribution: Record<string, DistributionData> = {};
+    const generators: Block[] = [];
 
     for (const send of this.sendQueue) {
+      // No race conditions! Verify we have the amount its trying to send still!
+      const currentStored = getMachineStorage(send.block, send.type);
+      const amountToSend = Math.min(currentStored, send.amount);
+
       if (send.type in distribution) {
         const data = distribution[send.type];
-        data.total += send.amount;
+        data.total += amountToSend;
         data.queueItems.push(send);
         continue;
       }
 
       distribution[send.type] = {
-        total: send.amount,
+        total: amountToSend,
         queueItems: [send],
       };
+
+      generators.push(send.block);
+
+      // Take away the amount generated:
+      // default behaviour of generate is to send the amount to generate + the current in the machine
+      // Instead do it by send.amount allowing for machines to use an alternative function to only output
+      // the amount generated, and not stored inside the machine always.
+      setMachineStorage(
```

Comment:
This seems like it could cause some weird issues. We don't know how long it's going to take until the last `sendGroupAllocation` is called. With the old system, the generators storage was set once rather than subtracting then waiting then adding again.